File: packages/typescript/spec/src/joins.ts
import { compareText } from "./canonical.js";
import type { GraphStructure } from "./types.js";

/** A connection belonging to an AND join, not an independently executing edge. */
export interface DerivedJoinEdge {
  joinId: string;
  source: string;
  target: string;
}

/**
 * Return fresh records from an already validated structure, sorted by joinId and
 * source in Unicode code point order without normalization. Validate the document
 * at the input boundary. Read both edges and joins; these links preserve AND
 * convergence and do not carry ordinary edge IDs/kinds or execution semantics.
 */
export function derivedJoinEdges(structure: GraphStructure): DerivedJoinEdge[] {
  return structure.joins
    .flatMap((join) =>
      join.sources.map((source) => ({
        joinId: join.id,
        source,
        target: join.target,
      })),
    )
    .sort(
      (left, right) =>
        compareText(left.joinId, right.joinId) ||
        compareText(left.source, right.source),
    );
}

File: packages/python/spec/src/agent_topology/spec/_joins.py
"""Consumer connections that retain multi-source join identity."""

from typing import Any


def derived_join_edges(structure: dict[str, Any]) -> list[dict[str, str]]:
    """Return fresh {joinId, source, target} records from a validated structure.

    Sort by joinId, then source, in Unicode code point order without normalization.
    Validate the containing document at the input boundary; this is not a validator.
    Read both edges and joins: these links retain AND convergence semantics and
    do not represent independent edge execution or ordinary edge IDs/kinds.
    """
    return sorted(
        (
            {"joinId": join["id"], "source": source, "target": join["target"]}
            for join in structure["joins"]
            for source in join["sources"]
        ),
        key=lambda link: (link["joinId"], link["source"]),
    )
